File: demos_code_J2_conception_objet/06_break_package_cycle/avant/src/com/acme/billing/BillingService.java
package com.acme.billing;
import com.acme.orders.OrderService;

public class BillingService {
    private OrderService orders;
    public void setOrders(OrderService orders) { this.orders = orders; }
    public void refund(String id) {
        System.out.println("billing: refund " + id);
        orders.markRefunded(id);
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/06_break_package_cycle/avant/src/com/acme/billing/billing_service.py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from com.acme.orders.order_service import OrderService


class BillingService:
    # Dépendance billing -> orders : le cycle architectural est visible.
    def __init__(self) -> None:
        self._orders: "OrderService | None" = None

    def set_orders(self, orders: "OrderService") -> None:
        self._orders = orders

    def refund(self, order_id: str) -> None:
        print(f"billing: refund {order_id}")
        if self._orders is None:
            raise RuntimeError("OrderService non configuré")
        self._orders.mark_refunded(order_id)
